man/install-mh.man
.TH INSTALL-MH %manext1% 2002-11-18 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
install-mh \- initialize the nmh environment
.SH SYNOPSIS
.HP 5
.na
.B %nmhlibexecdir%/install\-mh
.RB [ \-help ]
.RB [ \-version ]
.RB [ \-auto ]
.RB [ \-check ]
.ad
.SH DESCRIPTION
.B install\-mh
creates the initial setup for a first-time
.B nmh
user.
.B install\-mh
is installed in two locations for historical reasons.
.PP
The \fB\-auto\fP switch does things as automatically as possible and
makes \fBinstall\-mh\fP less chatty.
.PP
The user is prompted for the name of the directory to be
designated as the user's
.B nmh
directory.  If this directory does not exist, the user is asked if
it should be created.  Normally, this directory should be under
the user's home directory, and has the default name of
\fI`Mail'\fP.
.B install\-mh
writes an initial
.I \&.mh_profile
for the user.
.PP
As with all
.B nmh
commands,
.B install\-mh
first checks for the existence of the \fB$MH\fP environment variable,
since that, if set, gives the profile path.
If it isn't set, the
.B $HOME
environment variable is consulted to determine the user's home directory.
If
.B $HOME
is not set, then the
.I /etc/passwd
file is consulted.
.PP
When creating the user's
.IR \&.mh_profile ,
.B install\-mh
will check for the existence of a global profile at
.IR %nmhetcdir%/mh.profile .
If found, this will be used to initialize the new
.IR \&.mh_profile .
.PP
The \fB\-check\fP switch checks if
.B nmh
is installed.
(This can be used by other programs to determine if
.B nmh
has been installed.)
.SH FILES
.PD 0
.TP 20
$HOME/.mh_profile
The user's profile.
.TP
%nmhetcdir%/mh.profile
Used to initialize user's profile.
.PD
.SH "PROFILE COMPONENTS"
.TP 20
Path:
To set the user's nmh directory.
.SH CONTEXT
With
.BR \-auto ,
the current folder is changed to
.RI \*(lq inbox \*(rq.


NEWS
#
# NEWS - Release notes for nmh 1.6
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been nearly two years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

------------
NEW FEATURES
------------

The following are new features for the 1.6 release of nmh:

- All nmh programs will now warn about multiple profile entries for
  the same non-null, non-comment component.  In previous versions, all
  but the first were silently ignored.
- mhmail now supports all post(8) options, or optionally can use
  send(1) with all of its options.  Its formerly undocumented -resent
  switch has been documented.  And it has new -attach, -headerfield,
  and -send/-nosend switches.
- Default to flock() locking on OpenBSD and Darwin.
- Added -directives support to mhbuild(1) to control whether or not
  mhbuild will honor MIME directives by default.  And added support
  for special #on/#off/#pop directives to control the MIME directive
  processing state.
- Added -messageid switch to send(1) and post(8).  This allows
  selection of the style to use for generated Message-ID and
  Content-ID header fields.  The default localname style is
  pid.time@localname, where time is in seconds, and matches previous
  behavior.  The random style replaces the localname with some
  (pseudo)random bytes and uses microsecond-resolution time.
- Added -clobber switch to mhstore(1) to control overwriting of
  existing files.
- Added -outfile switch to mhstore(1).
- Added -noall/-all switches to sortm(1).  sortm -noall requires
  a messages argument.
- $PAGER overrides the compiled-in default pager command.
- Added etc/scan.highlighted format file, as an example of how to
  highlight/colorize the output of scan(1).
- inc(1) now supports a -port switch to specify the port used by the
  POP server.
- pick(1) now decodes MIME-encoded header fields before searching.
- The VISUAL and EDITOR environment variables are now supported as fallbacks
  if the user does not configure an editor entry in their profile.
- The format engine (mh_format(5)) now properly accounts for multibyte
  characters when accounting for column widths.
- burst(1) now can burst MIME-formatted digests (messages that contain
  message/rfc822 parts instead of messages formatted with RFC 934).
- All proc entries (showproc, moreproc, etc) can now accept entries that
  contain spaces and shell metacharacters.  If found, such entries will
  either be space-split or processed by /bin/sh.
- A new program, fmttest(1) is included to help debug format files.  See
  NOTABLE CHANGES below.
- mhshow/mhstore now have support for RFC-2017 (access-type=url) for
  external message bodies.
- Added -retainsequences switch to refile(1).
- A new program, mhfixmsg(1), is included to rewrite MIME messages with
  various transformations.  See NOTABLE CHANGES below.
- Added -[no]rmmproc switches to rmm(1).
- Added support for Content-Disposition header (RFC 2183) to mhstore(1)
  and mhn(1) when used with -auto.
- All nmh commands now support transactional locking for sequence files.
  See NOTABLE CHANGES below.
- There is no longer a per-folder maximum number of sequences.
- For the SMTP MTA TLS can now be negotiated at the beginning of the
  connection with the -initialtls switch.
- Messages can now be selected using a relative offset from some other
  message, or from the start or end of a sequence.  See mh-sequence(5).
- The -changecur and -nochangecur switches have been added to mhlist(1).
- mhbuild(1) can now encode 8-bit message headers using RFC-2047 encoding
  rules.  See NOTABLE CHANGES below.
- The whatnow(1) attach feature will determine the content type of an
  attachment using a program such as file --mime-type, if available at
  configuration time.  If not, it will use mhshow-suffix- entries as
  before.  The -v switch to attach causes it to display the mhbuild
  directive that send(1) will use.
- mhbuild(1) now supports the -auto/-noauto flags (to be used by send(1)
  when invoking mhbuild automatically).
- mhbuild(1) now is automatically run by send, to insure that all outgoing
  messages have proper MIME formatting.  See NOTABLE CHANGES below.
- A new header, "Attach", is supported by mhbuild; it is used to replace
  previous functionality (which by default used a header named
  Nmh-Attachment).  See NOTABLE CHANGES below.
- The default Content-Transfer-Encoding for text parts is now 8bit.
- mhbuild(1) now supports a selectable Content-Transfer-Encoding
- If nmh was configured with iconv(3) support, mhshow will convert, if
  necessary, the charset of text/plain content to match the user's
  locale setting.  See NOTABLE CHANGES below.
- Added support for %{charset} display string escape to mhshow(1).
- The MIME parsing and generating routines now support RFC 2231 extended
  parameter information.
- mh-mime(7) now provides an introduction to nmh's MIME handling.
- mhshow(1) will now by default display all text content under one pager,
  and display markers for non-text and non-inline content.  The content
  markers are changeable via mh-format(5).  See NOTABLE CHANGES below.

-----------------
OBSOLETE FEATURES
-----------------
- Changed exit status of each nmh command's -version and -help
  switches from 1 to 0.
- The following environment variables were deprecated in nmh 1.5
  and removed from this release:
    MHPOPDEBUG (use -snoop command line switch instead)
    MM_NOASK (use -nolist and -nopause command line switches instead)
    NOMHNPROC (use -nocheckmime command line switch instead)
    FACEPROC (undocumented faceproc feature removed)
- Changed repl and dist default to -noatfile.  The default of -atfile

uip/install-mh.c
/* install-mh.c -- initialize the nmh environment of a new user
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include <h/mh.h>				/* mh internals */
#include <h/utils.h>
#include <pwd.h>				/* structure for getpwuid() results */

#define INSTALLMH_SWITCHES \
    X("auto", 0, AUTOSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \
    X("check", 1, CHECKSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(INSTALLMH);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(INSTALLMH, switches);
#undef X


int
main (int argc, char **argv)
{
    int autof = 0;
    char *cp, buf[BUFSIZ];
    const char *pathname;
    char *dp, **arguments, **argp;
    struct node *np;
    struct passwd *pw;
    struct stat st;
    FILE *in, *out;
    int		check;

    if (nmh_init(argv[0], 0 /* use context_foil() */ )) { return 1; }

    arguments = getarguments (invo_name, argc, argv, 0);
    argp = arguments;

    check = 0;

    while ((dp = *argp++)) {
	if (*dp == '-') {
	    switch (smatch (++dp, switches)) {
		case AMBIGSW:
		    ambigsw (dp, switches);
		    done (1);
		case UNKWNSW:
		    adios (NULL, "-%s unknown\n", dp);

		case HELPSW:
		    snprintf (buf, sizeof(buf), "%s [switches]", invo_name);
		    print_help (buf, switches, 0);
		    done (0);
		case VERSIONSW:
		    print_version(invo_name);
		    done (0);

		case AUTOSW:
		    autof++;
		    continue;

		case CHECKSW:
		    check = 1;
		    continue;
	    }
	} else {
	    adios (NULL, "%s is invalid argument", dp);
	}
    }

    /*
     *	Find user's home directory.  Try the HOME environment variable first,
     *	the home directory field in the password file if that's not found.
     */

    if ((mypath = getenv("HOME")) == NULL) {
	if ((pw = getpwuid(getuid())) == NULL || *pw->pw_dir == '\0')
	    adios(NULL, "cannot determine your home directory");
	else
	    mypath = pw->pw_dir;
    }

    /*
     *	Find the user's profile.  Check for the existence of an MH environment
     *	variable first with non-empty contents.  Convert any relative path name
     *	found there to an absolute one.  Look for the profile in the user's home
     *	directory if the MH environment variable isn't set.
     */

    if ((cp = getenv("MH")) && *cp != '\0')
	defpath = path(cp, TFILE);
    else
	defpath = concat(mypath, "/", mh_profile, NULL);

    /*
     *	Check for the existence of the profile file.  It's an error if it exists and
     *	this isn't an installation check.  An installation check fails if it does not
     *	exist, succeeds if it does.
     */

    if (stat (defpath, &st) != NOTOK) {
	if (check)
	    done(0);

	else if (autof)
	    adios (NULL, "invocation error");
	else
	    adios (NULL, "You already have an nmh profile, use an editor to modify it");
    }
    else if (check) {
	done(1);
    }

    if (!autof && read_switch ("Do you want help? ", anoyes)) {
	(void)printf(
	 "\n"
	 "Prior to using nmh, it is necessary to have a file in your login\n"
	 "directory (%s) named %s which contains information\n"
	 "to direct certain nmh operations.  The only item which is required\n"
	 "is the path to use for all nmh folder operations.  The suggested nmh\n"
	 "path for you is %s/Mail...\n"
	 "\n", mypath, mh_profile, mypath);
    }

    cp = concat (mypath, "/", "Mail", NULL);
    if (stat (cp, &st) != NOTOK) {
	if (S_ISDIR(st.st_mode)) {
	    cp = concat ("You already have the standard nmh directory \"",
		    cp, "\".\nDo you want to use it for nmh? ", NULL);
	    if (read_switch (cp, anoyes))
		pathname = "Mail";
	    else
		goto query;
	} else {
	    goto query;
	}
    } else {
	if (autof)
	    puts("I'm going to create the standard nmh path for you.");
	else
	    cp = concat ("Do you want the standard nmh path \"",
		    mypath, "/", "Mail\"? ", NULL);
	if (autof || read_switch (cp, anoyes))
	    pathname = "Mail";
	else {
query:
	    if (read_switch ("Do you want a path below your login directory? ",
			anoyes)) {
		printf ("What is the path?  %s/", mypath);
		pathname = read_line ();
		if (pathname == NULL) done (1);
	    } else {
		printf ("What is the whole path?  /");
		pathname = read_line ();
		if (pathname == NULL) done (1);
		pathname = concat ("/", pathname, NULL);
	    }
	}
    }

    if (chdir (mypath) < 0) {
	advise (mypath, "chdir");
    }
    if (chdir (pathname) == NOTOK) {
	cp = concat ("\"", pathname, "\" doesn't exist; Create it? ", NULL);
	if (autof || read_switch (cp, anoyes))
	    if (makedir (pathname) == 0)
		adios (NULL, "unable to create %s", pathname);
    } else {
	puts("[Using existing directory]");
    }

    /*
     * Add some initial elements to the profile/context list
     */
    NEW(np);
    m_defs = np;
    np->n_name = mh_xstrdup("Path");
    np->n_field = mh_xstrdup(pathname);
    np->n_context = 0;
    np->n_next = NULL;

    /*
     * If there is a default profile file in the
     * nmh `etc' directory, then read it also.
     */
    if ((in = fopen (mh_defaults, "r"))) {
	readconfig (&np->n_next, in, mh_defaults, 0);
	fclose (in);
    }

    ctxpath = getcpy (m_maildir (context = "context"));

    /* Initialize current folder to default */
    context_replace (pfolder, defaultfolder);
    context_save ();

    /*
     * Now write out the initial .mh_profile
     */
    if ((out = fopen (defpath, "w")) == NULL)
	adios (defpath, "unable to write");
    /*
     * The main purpose of this first line is to fool file(1).
     * Without it, if the first line of the profile is Path:,
     * file 5.19 reports its type as message/news.  With it,
     * it reports the type as text/plain.
     */
    fprintf (out, "MH-Profile-Version: 1.0\n");
    for (np = m_defs; np; np = np->n_next) {
	if (!np->n_context)
	    fprintf (out, "%s: %s\n", np->n_name, np->n_field);
    }
    fclose (out);

    puts ("\nPlease see the nmh(7) man page for an introduction to nmh.\n");
    print_intro (stdout, FALSE);

    /* Initialize the saved nmh version.  The Path profile entry was added
       above, that's all this needs. */
    (void) nmh_version_changed (0);

    done (0);
    return 1;
}
